src/worker/jobs/send_publish_notifications.rs

```diff
@@ -0,0 +1,162 @@
+use crate::email::Email;
+use crate::models::OwnerKind;
+use crate::schema::{crate_owners, crates, emails, users, versions};
+use crate::tasks::spawn_blocking;
+use crate::worker::Environment;
+use anyhow::anyhow;
+use chrono::{NaiveDateTime, SecondsFormat};
+use crates_io_worker::BackgroundJob;
+use diesel::prelude::*;
+use diesel_async::{AsyncPgConnection, RunQueryDsl};
+use std::sync::Arc;
+
+/// Background job that sends email notifications to all crate owners when a
+/// new crate version is published.
+#[derive(Serialize, Deserialize)]
+pub struct SendPublishNotificationsJob {
+    version_id: i32,
+}
+
+impl SendPublishNotificationsJob {
+    pub fn new(version_id: i32) -> Self {
+        Self { version_id }
+    }
+}
+
+impl BackgroundJob for SendPublishNotificationsJob {
+    const JOB_NAME: &'static str = "send_publish_notifications";
+
+    type Context = Arc<Environment>;
+
+    async fn run(&self, ctx: Self::Context) -> anyhow::Result<()> {
+        let mut conn = ctx.deadpool.get().await?;
+
+        // Get crate name, version and other publish details
+        let publish_details = PublishDetails::for_version(self.version_id, &mut conn).await?;
+
+        let publish_time = publish_details
+            .publish_time
+            .and_utc()
+            .to_rfc3339_opts(SecondsFormat::Secs, true);
+
+        // Find names and email addresses of all crate owners
+        let recipients = crate_owners::table
+            .filter(crate_owners::deleted.eq(false))
+            .filter(crate_owners::owner_kind.eq(OwnerKind::User))
```

Comment:
Should we also be looking up team owners?

Reply:
we should, I'm just not sure that we can... :D

our publishing code currently performs a "get all owner teams, iterate through teams and query GitHub API to see if publisher is member of the team" operation, but that does not list the members of these teams. in other words: we'd need new GitHub API calls since we don't "cache" the team memberships on our side. if/when we have non-GitHub teams implemented this will probably become a lot easier.

Reply:
Ugh, I forgot that we don't cache that. Sorry, some Sourcegraph slipped through there.

Not blocking for this PR, but I'm wondering if it's worth the effort to figure that[^that] out sooner rather than later — given that part of the motivation here (at least from what I see) is supply chain security, not sending notifications for crates that are entirely owned by teams (such as the AWS crate ecosystem) feels problematic.

I'll give this some thought.

[^that]: For clarity, I'm more thinking just some level of caching here; not so much a full blown decoupled-from-GitHub team design.

Reply:
> I'm wondering if it's worth the effort to figure that out sooner rather than later

it then becomes a problem of cache invalidation. I think I'd personally rather see a proper teams concept in crates.io with a feature to optionally keep the team memberships in sync with some GitHub team. Pietro mentioned a while ago that this sync might be possible by having people install a GitHub app (?) in their orgs, but I haven't looked at any of this in detail yet.